## Dataset Information:
Each column provides specific information about the patient, their admission, and the healthcare services provided, making this dataset suitable for various data analysis and modeling tasks in the healthcare domain. Here's a brief explanation of each column in the dataset  

Name: This column represents the name of the patient associated with the healthcare record.  
Age: The age of the patient at the time of admission, expressed in years.  
Gender: Indicates the gender of the patient, either "Male" or "Female."  
Blood Type: The patient's blood type, which can be one of the common blood types (e.g., "A+", "O-", etc.).  
Medical Condition: This column specifies the primary medical condition or diagnosis associated with the patient, such as "Diabetes," "Hypertension," "Asthma," and more.  
Date of Admission: The date on which the patient was admitted to the healthcare facility.  
Doctor: The name of the doctor responsible for the patient's care during their admission.  
Hospital: Identifies the healthcare facility or hospital where the patient was admitted.  
Insurance Provider: This column indicates the patient's insurance provider, which can be one of several options, including "Aetna," "Blue Cross," "Cigna," "UnitedHealthcare," and "Medicare."  
Billing Amount: The amount of money billed for the patient's healthcare services during their admission. This is expressed as a floating-point number.  
Room Number: The room number where the patient was accommodated during their admission.  
Admission Type: Specifies the type of admission, which can be "Emergency," "Elective," or "Urgent," reflecting the circumstances of the admission.  
Discharge Date: The date on which the patient was discharged from the healthcare facility, based on the admission date and a random number of days within a realistic range.  
Medication: Identifies a medication prescribed or administered to the patient during their admission. Examples include "Aspirin," "Ibuprofen," "Penicillin," "Paracetamol," and "Lipitor."  
Test Results: Describes the results of a medical test conducted during the patient's admission. Possible values include "Normal," "Abnormal," or "Inconclusive," indicating the outcome of the test.  

The dataset covers the period from 2019-05-08 to 2024-06-06

## Import libraries

In [1]:
import numpy as np
import pandas as pd

## Read the data from the CSV file

In [2]:
df = pd.read_csv('data/healthcare_dataset.csv')

In [3]:
df.head(5)

Name  Age  Gender Blood Type Medical Condition Date of Admission  \
0  Bobby JacksOn   30    Male         B-            Cancer        2024-01-31   
1   LesLie TErRy   62    Male         A+           Obesity        2019-08-20   
2    DaNnY sMitH   76  Female         A-           Obesity        2022-09-22   
3   andrEw waTtS   28  Female         O+          Diabetes        2020-11-18   
4  adrIENNE bEll   43  Female        AB+            Cancer        2022-09-19   

             Doctor                    Hospital Insurance Provider  \
0     Matthew Smith             Sons and Miller         Blue Cross   
1   Samantha Davies                     Kim Inc           Medicare   
2  Tiffany Mitchell                    Cook PLC              Aetna   
3       Kevin Wells  Hernandez Rogers and Vang,           Medicare   
4    Kathleen Hanna                 White-White              Aetna   

   Billing Amount  Room Number Admission Type Discharge Date   Medication  \
0    18856.281306          328         Urgent     2024-02-02  Paracetamol   
1    33643.327287          265      Emergency     2019-08-26    Ibuprofen   
2    27955.096079          205      Emergency     2022-10-07      Aspirin   
3    37909.782410          450       Elective     2020-12-18    Ibuprofen   
4    14238.317814          458         Urgent     2022-10-09   Penicillin   

   Test Results  
0        Normal  
1  Inconclusive  
2        Normal  
3      Abnormal  
4      Abnormal

In [4]:
df.shape

(55500, 15)

### The dataset has 55500 rows and 15 columns

In [5]:
df.dtypes

Name                   object
Age                     int64
Gender                 object
Blood Type             object
Medical Condition      object
Date of Admission      object
Doctor                 object
Hospital               object
Insurance Provider     object
Billing Amount        float64
Room Number             int64
Admission Type         object
Discharge Date         object
Medication             object
Test Results           object
dtype: object

### Some columns need to change the data type (Date of Admission and Discharge Date to datetime type)

In [6]:
df.isna().any()

Name                  False
Age                   False
Gender                False
Blood Type            False
Medical Condition     False
Date of Admission     False
Doctor                False
Hospital              False
Insurance Provider    False
Billing Amount        False
Room Number           False
Admission Type        False
Discharge Date        False
Medication            False
Test Results          False
dtype: bool

### The dataset does not contain missing data

## Changing data type for Date of Admission and Discharge Date columns

In [7]:
for col in df.columns:
    if 'date' in col.lower():
        df[col] = pd.to_datetime(df[col])


## Bring the names of patients to a unified form

In [8]:
df['Name'] = df['Name'].str.title()

In [9]:
df.head(3)

Name  Age  Gender Blood Type Medical Condition Date of Admission  \
0  Bobby Jackson   30    Male         B-            Cancer        2024-01-31   
1   Leslie Terry   62    Male         A+           Obesity        2019-08-20   
2    Danny Smith   76  Female         A-           Obesity        2022-09-22   

             Doctor         Hospital Insurance Provider  Billing Amount  \
0     Matthew Smith  Sons and Miller         Blue Cross    18856.281306   
1   Samantha Davies          Kim Inc           Medicare    33643.327287   
2  Tiffany Mitchell         Cook PLC              Aetna    27955.096079   

   Room Number Admission Type Discharge Date   Medication  Test Results  
0          328         Urgent     2024-02-02  Paracetamol        Normal  
1          265      Emergency     2019-08-26    Ibuprofen  Inconclusive  
2          205      Emergency     2022-10-07      Aspirin        Normal

## Duplicates checking

In [10]:
df.duplicated().value_counts()

False    54966
True       534
Name: count, dtype: int64

### The dataset has 534 duplicates that should be removed before further analysis

## Description of columns with numerical values

In [11]:
df['Billing Amount'].describe()

count    55500.000000
mean     25539.316097
std      14211.454431
min      -2008.492140
25%      13241.224652
50%      25538.069376
75%      37820.508436
max      52764.276736
Name: Billing Amount, dtype: float64

## The largest amount for treatment is 52764 dollars, the smallest amount is -2008 dollars.  
## Average check: 25,539 dollars.  

Since we have negative values, we have to check the origin of such values ​​- is this the norm (return of the part reimbursed by the insurance company), or is it an error in entering data into the database.
To do this, we will sort the data exclusively with negative dates of payment for treatment, and then we will check randomly on an individual patient whether there is only one record or on the same dates, with the same diagnosis, but with a positive value of the amount to be paid.

In [12]:
df[df['Billing Amount'] < 0].sort_values(by='Name').head(10)

Name  Age  Gender Blood Type Medical Condition  \
48994      Aaron Flowers   42    Male         A+            Asthma   
8757          Alan Scott   48  Female         A-            Asthma   
6798   Alexander Gardner   45    Male        AB-            Cancer   
34304   Alexander Martin   46    Male         A+      Hypertension   
5445      Alexandra Khan   32    Male        AB+         Arthritis   
43101      Amanda Morgan   35  Female        AB+      Hypertension   
6961        Angela Allen   83  Female         B-          Diabetes   
17642     Anthony Miller   50  Female         O+         Arthritis   
132      Ashley Erickson   32  Female        AB-            Cancer   
15542      Ashley Ramsey   53    Male        AB+            Cancer   

      Date of Admission            Doctor                     Hospital  \
48994        2020-06-03     Ronald Patton     Brown, Santos and Howard   
8757         2019-10-23       Jacob Kirby                 Lopez-Hughes   
6798         2023-03-23       Tracy Brown              Maldonado Group   
34304        2021-09-24       Tracy Smith      Carter, and Nguyen Snow   
5445         2022-07-14    Michael Vaughn       Bowen Lopez, and Terry   
43101        2020-09-27  Gregory Figueroa           Petersen-Hernandez   
6961         2019-08-21    Robert Kennedy    Brown and Randolph, Young   
17642        2020-01-17      Jill Raymond  Pitts, Hernandez and Wilson   
132          2019-11-05     Gerald Hooper    and Johnson Moore, Branch   
15542        2019-05-30       James Smith              Barker-Mitchell   

      Insurance Provider  Billing Amount  Room Number Admission Type  \
48994           Medicare     -378.696411          115       Elective   
8757            Medicare     -857.125933          309         Urgent   
6798               Cigna     -198.283806          329       Elective   
34304   UnitedHealthcare      -38.966152          401      Emergency   
5445               Aetna     -692.408820          372       Elective   
43101           Medicare     -652.181369          144       Elective   
6961               Cigna      -43.098522          189         Urgent   
17642              Cigna     -892.270140          359      Emergency   
132                Aetna     -502.507813          376         Urgent   
15542         Blue Cross     -135.719073          235       Elective   

      Discharge Date   Medication  Test Results  
48994     2020-06-17  Paracetamol  Inconclusive  
8757      2019-11-18  Paracetamol      Abnormal  
6798      2023-04-09      Aspirin      Abnormal  
34304     2021-09-26      Lipitor  Inconclusive  
5445      2022-07-19      Lipitor      Abnormal  
43101     2020-10-20   Penicillin      Abnormal  
6961      2019-08-24   Penicillin      Abnormal  
17642     2020-01-31      Aspirin      Abnormal  
132       2019-11-23   Penicillin        Normal  
15542     2019-06-09   Penicillin  Inconclusive

In [13]:
df[df['Name'] == 'Ashley Ramsey']

Name  Age Gender Blood Type Medical Condition  \
15542  Ashley Ramsey   53   Male        AB+            Cancer   

      Date of Admission       Doctor         Hospital Insurance Provider  \
15542        2019-05-30  James Smith  Barker-Mitchell         Blue Cross   

       Billing Amount  Room Number Admission Type Discharge Date  Medication  \
15542     -135.719073          235       Elective     2019-06-09  Penicillin   

       Test Results  
15542  Inconclusive

### According to the conducted analysis, we can state that the negative values ​​of payment amounts for treatment are an error in entering the database and such terms are subject to deletion before further analysis.

## Creating clean dataset

In [14]:
clean_data = df[df['Billing Amount'] > 0].copy(deep=True)
clean_data.drop_duplicates(inplace=True)

In [15]:
clean_data.head(3)

Name  Age  Gender Blood Type Medical Condition Date of Admission  \
0  Bobby Jackson   30    Male         B-            Cancer        2024-01-31   
1   Leslie Terry   62    Male         A+           Obesity        2019-08-20   
2    Danny Smith   76  Female         A-           Obesity        2022-09-22   

             Doctor         Hospital Insurance Provider  Billing Amount  \
0     Matthew Smith  Sons and Miller         Blue Cross    18856.281306   
1   Samantha Davies          Kim Inc           Medicare    33643.327287   
2  Tiffany Mitchell         Cook PLC              Aetna    27955.096079   

   Room Number Admission Type Discharge Date   Medication  Test Results  
0          328         Urgent     2024-02-02  Paracetamol        Normal  
1          265      Emergency     2019-08-26    Ibuprofen  Inconclusive  
2          205      Emergency     2022-10-07      Aspirin        Normal

### Continue to explore numerical values

In [16]:
clean_data[['Age', 'Billing Amount']].describe()

Age  Billing Amount
count  54860.000000    54860.000000
mean      51.533850    25594.633637
std       19.605295    14175.867041
min       13.000000        9.238787
25%       35.000000    13299.747940
50%       52.000000    25593.873000
75%       68.000000    37847.066671
max       89.000000    52764.276736

### After data cleaning and validation, we have the following values:
max Billing Amount is 52764.27 dollars, min Billing Amount is 9.23 dollars.  
Average check: 25594.63 dollars.  

The distribution of the age indicator varies from a minimum of 13 years to a maximum of 89 years. The average age of patients is 51.5 years.  

In [17]:
clean_data[['Gender']].value_counts()

Gender
Male      27449
Female    27411
Name: count, dtype: int64

In [18]:
clean_data[['Test Results']].value_counts()

Test Results
Abnormal        18399
Normal          18302
Inconclusive    18159
Name: count, dtype: int64

### The distribution by Gender and Test Results is even

In [19]:
clean_data[['Medical Condition']].value_counts()

Medical Condition
Arthritis            9207
Diabetes             9197
Hypertension         9131
Obesity              9127
Cancer               9121
Asthma               9077
Name: count, dtype: int64

In [20]:
clean_data['Year of Admission'] = clean_data['Date of Admission'].dt.year
clean_data['Month of Admission'] = clean_data['Date of Admission'].dt.month

In [21]:
clean_data['Year of Admission'].value_counts(sort='True')

Year of Admission
2020    11157
2023    10918
2022    10891
2021    10795
2019     7280
2024     3819
Name: count, dtype: int64

In [22]:
clean_data.groupby('Month of Admission')['Medical Condition'].count()

Month of Admission
1     4646
2     4199
3     4610
4     4466
5     4547
6     4640
7     4762
8     4777
9     4503
10    4605
11    4501
12    4604
Name: Medical Condition, dtype: int64

In [23]:
clean_data.groupby('Medical Condition')['Billing Amount'].agg(['count', 'mean', 'min', 'max'])

count          mean        min           max
Medical Condition                                              
Arthritis           9207  25542.898356  32.626843  52170.036854
Asthma              9077  25685.385020  42.514589  52181.837792
Cancer              9121  25205.924207   9.238787  52373.032374
Diabetes            9197  25714.326349  31.030955  52211.852966
Hypertension        9131  25559.844868  68.909747  52764.276736
Obesity             9127  25859.215373  53.931445  52024.726443

In [24]:
clean_data.groupby('Blood Type')['Medical Condition'].count()

Blood Type
A+     6879
A-     6882
AB+    6872
AB-    6859
B+     6876
B-     6861
O+     6843
O-     6788
Name: Medical Condition, dtype: int64

In [25]:
clean_data['Age'].value_counts()

Age
38    890
37    880
57    876
34    858
80    854
     ... 
88     25
16     24
14     18
13     14
89      8
Name: count, Length: 77, dtype: int64

In [26]:
min_age = clean_data['Age'].min()
max_age = clean_data['Age'].max()

In [27]:
bins = pd.interval_range(start=min_age, end=max_age, freq=(max_age - min_age) // 10, closed='left')

In [28]:
clean_data['Age bins'] = pd.cut(clean_data['Age'], bins=bins, include_lowest=True)

In [29]:
clean_data.head(3)

Name  Age  Gender Blood Type Medical Condition Date of Admission  \
0  Bobby Jackson   30    Male         B-            Cancer        2024-01-31   
1   Leslie Terry   62    Male         A+           Obesity        2019-08-20   
2    Danny Smith   76  Female         A-           Obesity        2022-09-22   

             Doctor         Hospital Insurance Provider  Billing Amount  \
0     Matthew Smith  Sons and Miller         Blue Cross    18856.281306   
1   Samantha Davies          Kim Inc           Medicare    33643.327287   
2  Tiffany Mitchell         Cook PLC              Aetna    27955.096079   

   Room Number Admission Type Discharge Date   Medication  Test Results  \
0          328         Urgent     2024-02-02  Paracetamol        Normal   
1          265      Emergency     2019-08-26    Ibuprofen  Inconclusive   
2          205      Emergency     2022-10-07      Aspirin        Normal   

   Year of Admission  Month of Admission  Age bins  
0               2024                   1  [27, 34)  
1               2019                   8  [62, 69)  
2               2022                   9  [76, 83)

In [30]:
clean_data['Age bins'].value_counts()

Age bins
[55, 62)    5720
[62, 69)    5714
[34, 41)    5706
[48, 55)    5690
[76, 83)    5668
[41, 48)    5667
[20, 27)    5607
[69, 76)    5591
[27, 34)    5435
[13, 20)    1674
Name: count, dtype: int64

In [31]:
clean_data['Age bins'] = clean_data['Age bins'].astype(str)

In [32]:
clean_data[clean_data['Age bins'] == '[13, 20)']['Medical Condition'].value_counts()

Medical Condition
Arthritis       295
Cancer          282
Obesity         276
Asthma          276
Hypertension    276
Diabetes        269
Name: count, dtype: int64

In [33]:
clean_data[clean_data['Billing Amount'] > 50000]['Medical Condition'].value_counts()

Medical Condition
Arthritis       92
Hypertension    80
Cancer          79
Diabetes        71
Asthma          66
Obesity         66
Name: count, dtype: int64